artman/cli/main2.py

```diff
@@ -0,0 +1,435 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""The new artman CLI with the following syntax.
+
+    artman [Options] generate|publish <artifact_name>
+
+Note: Only local execution is supported as this moment. The CLI syntax is beta,
+and might have changes in the future.
+"""
+
+from __future__ import absolute_import
+from logging import DEBUG, INFO
+import argparse
+import io
+import os
+import pprint
+import subprocess
+import sys
+
+from ruamel import yaml
+
+from taskflow import engines
+
+from artman.config import converter, loader
+from artman.config.proto import config_pb2
+from artman.cli import support
+from artman.pipelines import pipeline_factory
+from artman.utils import config_util
+from artman.utils.logger import logger, setup_logging
+
+ARTMAN_DOCKER_IMAGE = 'googleapis/artman:0.4.12'
+
+
+def main(*args):
+    """Main method of artman.
+
+    Note: artman currently won't work if the input directory doesn't contain
+    some shared configuration file (e.g. gapic/packaging/dependencies.yaml).
+    This will make artman less useful for non-Google APIs.
+    TODO(ethanbao): Fix that by checking the input directory and downloading
+    the common yamls if necessary.
+    """
+    # If no arguments are sent, we are using the entry point; derive
+    # them from sys.argv.
+    if not args:
+        args = sys.argv[1:]
+
+    # Get to a normalized set of arguments.
+    flags = parse_args(*args)
+    user_config = read_user_config(flags)
+    pipeline_name, pipeline_kwargs = normalize_flags(flags, user_config)
+
+    if flags.local:
+        pipeline = pipeline_factory.make_pipeline(pipeline_name, False,
+                                                  **pipeline_kwargs)
+        # Hardcoded to run pipeline in serial engine, though not necessarily.
+        engine = engines.load(
+            pipeline.flow, engine='serial', store=pipeline.kwargs)
+        engine.run()
+        _chown_for_artman_output(os.path.abspath(flags.output_dir))
+    else:
+        support.check_docker_requirements(flags.image)
+        logger.info('Running artman command in a Docker instance.')
+        _run_artman_in_docker(flags)
+
+
+def parse_args(*args):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--config',
+        type=str,
+        default='artman.yaml',
+        help='[Optional] Specify path to artman config yaml, which can be '
+        'either an absolute path, or a path relative to the input '
+        'directory (specified by `--input-dir` flag). Default to '
+        '`artman.yaml`', )
```

Comment:
I have mixed feeling. I don't like such magic for any input file, but it is also not too controversial if you see it as BUILD file of bazel, setup.py for python, or pom.xml for maven.